### PR TITLE
ATLAS-126: Created RSS feed of updates to Atlas data

### PR DESCRIPTION
**Pull request by HelioStrike**
JIRA ticket: https://issues.openmrs.org/browse/ATLAS-126

I'm not entirely sure what an RSS feed should look like, and what some of those terms mean, so I will need some help here with the schema. :|

I created an SQL table called rss, seeing how some blogs (shared by @cintiadr !) used the database to store the feed entries.
Links to said blogs: https://davidwalsh.name/rss-express, http://programmerblog.net/how-to-generate-rss-feed-using-nodejs/

Here, every call to GET '/rss' uses the database and generates the feed again. Would it be better to use some 'last_cached' variable, so that we can create the feed only once in a given period?

[Here's](https://github.com/dylang/node-rss/blob/master/readme.md) the readme of the module I used, and the generated RSS feed. 

Screenshot from 2019-07-07 23-50-18

### CODE REVIEW

File: routes/rss.js [hunk 1]
```diff
@@ -0,0 +1,46 @@
+var express = require('express');
+var router = express.Router();
+var RSS = require('rss');
+
+module.exports = function(connection) {
+
+    const ATLAS_TITLE = 'OpenMRS Atlas';
```
Comment (bmamlin) [inline routes/rss.js]: `const ATLAS_TITLE = process.env.ATLAS_RSS_TITLE || 'OpenMRS Atlas';`

File: routes/rss.js [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+    const ATLAS_DESC = 'RSS feed for OpenMRS Atlas';
```
Comment (bmamlin) [inline routes/rss.js]: `const ATLAS_DESC = process.env.ATLAS_RSS_DESC || 'Updates to OpenMRS Atlas';`

File: routes/rss.js [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+    const ATLAS_LINK = 'https://atlas.openmrs.org';
```
Comment (bmamlin) [inline routes/rss.js]: `const ATLAS_LINK = process.env.ATLAS_RSS_LINK || 'https://atlas.openmrs.org';`

File: routes/rss.js [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+    const RSS_LINK = ATLAS_LINK + '/rss';
+    const FEED_LENGTH = 20;
```
Comment (bmamlin) [inline routes/rss.js]: `const FEED_LENGTH = process.env.ATLAS_RSS_LENGTH || 20;`

File: db/rsstable.sql [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+/*!40101 SET @saved_cs_client     = @@character_set_client */;
+/*!40101 SET character_set_client = utf8 */;
+CREATE TABLE `rss` (
+  `id` INT(10) unsigned NOT NULL AUTO_INCREMENT,
+  `title` VARCHAR(1024) NOT NULL,
+  `description` VARCHAR(4096) NOT NULL,
+  `author` VARCHAR(1024) NOT NULL,
+  `date` TIMESTAMP DEFAULT CURRENT_TIMESTAMP,
+  PRIMARY KEY (`id`)
+) ENGINE=InnoDB DEFAULT CHARSET=utf8 COLLATE=utf8_unicode_ci;
```
Comment (bmamlin) [inline db/rsstable.sql]: Including an index on `date` would make reads much faster.

**Comment by bmamlin** [comment]
If we could support a link to a specific marker (e.g., `/?marker={ID}`) that would open up Atlas with the given marker centered and its bubble opened, that would be the perfect link to include for an RSS entry. So, someone reading the feed would see "Foobar site joins the OpenMRS Atlas!", click the link, and be looking at the Foobar site's marker.

File: routes/markers.js [hunk 6]
```diff
@@ -308,6 +311,7 @@ module.exports = function(connection) {
                     else {
                         res.setHeader('Content-Type', 'application/json');
                         res.json(data);
+                        utils.addRSS(connection, 'Marker Deleted', `OpenMRS Atlas bids farewell to ${data.name}`, req.session.user.uid);
```
Comment (bmamlin) [inline routes/markers.js]: * Title: OpenMRS Atlas bids farewell to ${data.name}
* Description: ${username} removed ${data.name} from the OpenMRS Atlas.

The link for this one would, obviously, not point to the marker; rather, just be to https://atlas.openmrs.org/.

File: routes/markers.js [hunk 7]
```diff
@@ -249,6 +251,7 @@ module.exports = function(connection) {
                         else {
                             res.setHeader('Content-Type', 'application/json');                    
                             res.json(data);        
+                            utils.addRSS(connection, 'Marker Updated', `${data.name} updated on OpenMRS Atlas`, req.session.user.uid);
```
Comment (bmamlin) [inline routes/markers.js]: * Title: ${data.name} entry updated
* Description: ${username} made changes to ${data.name} on the OpenMRS Atlas

File: utils.js [hunk 8]
```diff
@@ -16,6 +16,16 @@ module.exports = {
         } else {
             res.send(401);
         }
-    }
+    },
     
+    /* Add entry to RSS table */
+    addRSS: function(connection, title, description, created_by) {
+
+        connection.query('INSERT INTO rss(title, description, author) VALUES(?,?,?)', [title, description, created_by], function (error, rows,field) {
```
Comment (bmamlin) [inline utils.js]: We might want to introduce a rate limiter or prevent duplicates (e.g., if someone hits the update link multiple times). Something like this (this is just pseudocode off the top of my head... you can probably come up with a better approach):
```node
if (recentlySeen.indexOf(title)) return; // ignore repeated entries
recentlySeen.push(title);
if (recentlySeen.length > 3) recentlySeen.shift();
```
A better approach might be something that prevents a duplicate for ~5 minutes or something like that.

**Comment by HelioStrike** [comment]
@bmamlin I implemented the requested changes. When we click on a marker link I wanted to not only open the marker window, but also set it as center and zoom in, but it doesn't seem to be working. It zooms in for a split-second, and then zooms out to cover the whole map. I couldn't find the code responsible for it, but I'll keep looking.

**Comment by bmamlin** [comment]
@HelioStrike, if you see the desired behavior but then the map "resets" to full screen, it's possible your code is working, but is executing before Google Maps has finished setting up... so Google Maps goes to its default state _after_ you've zoomed in.

Try delaying your code – e.g., `setTimeout(focusOnMarker, 5000)` – and see if it works properly after the delay. If that works, then what you need to find is the event/notification that Google Maps has finished loading and execute your code when that event/trigger fires. I found [this tip](https://stackoverflow.com/a/7262773/5602641), which might help.

**Pull request merged**